Updated Tentacle Resource

dodgydorito opened the pull request:
Completed the following Updates:
- Revamped tentacle registration so you can select which NIC to bind to. Useful for customer who want to pin OctopusDeploy traffic to a private network.
- Pulled project deployment into it's own resource. Useful for customers who want to deploy multiple projects at specific version numbers.

dodgydorito committed:
Modifications made to enable Tentacle registration based on user selected network adapter

dodgydorito committed:
Adding deploy project resource to abstract project deployment from tentacle deployment

dodgydorito committed:
Adding saved changes

dodgydorito committed:
Broke out project deploy into it's own resource

dodgydorito committed:
Updated ReadMe with new instructions on how to use the new resource module cProjectDeploy

dodgydorito committed:
Fixing Set-TargetResource to use DeployVersion only if it isn't latest.

nick-o commented on DSCResources/cTentacleAgent/cTentacleAgent.psm1:
```diff
@@ -244,11 +247,12 @@ function Invoke-AndAssert {
-    $downloader = new-object System.Net.WebClient
-    $ip = $downloader.DownloadString("http://icanhazip.com").Trim()
+    Write-Verbose "Getting IP address of $($RegisteredNic) NIC"
+    $ip = Get-NetIPAddress -InterfaceAlias $RegisteredNic -AddressFamily IPv4 | select -exp IPAddress
```
I can think of scenarios where this would not actually yield the desired public IP address. For instance a server behind a NAT (think VM in dedicated) with the Octopus server being unable to reach the tentacle via private IP (e.g. Octopus server located on customer's premise without VPN). I remember struggling with the same problem when porting this module for my customer and not entirely sure how to fix. Maybe add a switch that indicates whether we want NATed or private/adapter address?

nick-o commented on DSCResources/cProjectDeploy/cProjectDeploy.psm1:
```diff
@@ -0,0 +1,102 @@
+        if ($Wait){
+            $deployArguments += "--waitfordeployment"
+        }
+        Invoke-AndAssert { & .\octo.exe $deployArguments}
```
Invoke-AndAssert is a function in the cTentacleAgent resource-module so will not exist in the execution context of this module hence will break the Set-TargetResource at this point. Probably easiest to just copy that function over to here.

dodgydorito replied:
Fixed this with commit #7

Commit:
ProjectDeploy fix with Add Invoke-AndAssert function

dodgydorito committed:
Removed commented code and updated the IP Address function to allow DSC to register the tentacle with your selected network adapter.

dodgydorito commented:
Fixed Public IP Address function with:

https://github.com/bvandermark/OctopusDSC/blob/master/DSCResources/cTentacleAgent/cTentacleAgent.psm1#L231-L255

Some changes include making sure the selected Adapter actually exists, an isNatted param that will allow the resource to fallback to retrieving the public IP of the server from an external site, and finally the function will run a test tcp connection to the Octopus Server and will throw an error in the event that it can't reach it.

dodgydorito committed:
Bug Fixes:
-ProjectDeploy
*Added Request-File function to ProjectDeploy so it can download the deploy tools.
*Removed Name Parameter as it is not used.
*Removed Roles and DefaultAppDir parameters from schema as they are not used.

-TentacleAgent
*Found that Test-NetConnection does not work with URLs so I added a check that will remove the http:// part of a URL

dodgydorito committed:
Updated ReadMe to describe changes
